File: backend/src/models/Donor.js
import mongoose from "mongoose";

const donorSchema = new mongoose.Schema({
  name: { type: String, required: true },
  dob: { type: Date, required: true },
  weight: { type: Number, required: true },
  bloodType: { type: String, required: true },
  contact: { type: String, required: true },
  address: { type: String, required: true },
  district: { type: String, required: true },
  lat: { type: Number },
  lng: { type: Number },
  createdAt: { type: Date, default: Date.now },
});

const Donor = mongoose.model("Donor", donorSchema);

export default Donor;

File: backend/models/donor.py
from mongoengine import Document, StringField, DateTimeField, signals
from datetime import datetime
import pytz

class Donor(Document):
    name = StringField(required=True)
    blood_type = StringField()
    contact = StringField()
    address = StringField()
    district = StringField()
    weight = StringField()
    timeanddate = DateTimeField()  # Automatically updated on save

    @classmethod
    def pre_save(cls, sender, document, **kwargs):
        """Update the timeanddate field before saving with Indian timezone."""
        india_tz = pytz.timezone('Asia/Kolkata')
        document.timeanddate = datetime.now(india_tz)

    def to_json(self):
        return {
            "id": str(self.id),
            "name": self.name,
            "blood_type": self.blood_type,
            "contact": self.contact,
            "address": self.address,
            "district": self.district,
            "weight": self.weight,
            "timeanddate": self.timeanddate.isoformat() if self.timeanddate else None
        }

signals.pre_save.connect(Donor.pre_save, sender=Donor)